### ISSUE TITLE
Contribution offer for multi user support

### ISSUE THREAD

**Issue by Varstahl** [zadam/trilium#2143]
My team uses Trilium notes extensively, and the one issue they keep having is the lack of multi user support.

I've read your understandable stance in the [FAQ](https://github.com/zadam/trilium/wiki/FAQ#multi-user-support) and #169, and I agree that it's a whole different can of worms. Which is why we would like to contribute this feature to your app.

We would like to accept your guidelines or hints on how to approach this, while we will handle all the ACL, security, and other development tasks that this feature needs, including related future bug fixes. This should allow you to focus on trilium's further development, while enabling note sharing/editing on teams.

Is it an acceptable proposal?

**Comment by zadam** [comment]
Hi, first of all, thank you for the offer. But to be completely honest with you, I don't see myself accepting such changes in the next year or so. My current priority is filling in last remaining basic/important features (like in-document search), stabilize the application and release version 1.0. Even just accepting / merging such huge feature would be probably too disrupting/destabilizing to this goal.

If you really want to go that way, then I would encourage you to fork the project and work on it on your own. My stance on this is not necessarily final, I might revisit it in one (or rather 2) years. But for me the right time did not come yet.

**Comment by Varstahl** [comment]
Fair enough, we'll try to maintain a parallel version in this case, and eventually upstream if/when possible.

**Comment by kakaroto** [comment]
Thank you @zadam for the consideration and your answer.
I'm part of the team with @Varstahl and we hope to start working on those changes soon. I haven't yet fully decided on how to go forward, but I do want this to be as non-disruptive as possible. We only have had a cursory look at how trillium works internally, so the design is not yet formed, but if you have any advice or suggestions (as I suppose you've had a lot more time to brainstorm the idea), we'd appreciate it.

I think I want us to do this in multiple phases, with increasing level of complexity. 
The very first use case to implement would be to still have one account with be ability to lock notes/subtree with specific passwords. Right now, when I share my trillium notes with my team, by sharing the password, I can't lock any of the content from them because the encryption password is the same as the login password, so it's not useful. Being able to lock notes with a custom password would greatly improve the privacy concerns and we could have something as simple as one subtree per user as a very sketchy/temporary sort of multi user support. So a user would login to the 'global account' with a shared username/password, then would unlock their own subtree of notes with their personal password. (This seems to be the same idea as #1903)

I think the next step might be to have Trillium automatically try to open `$TRILLUM_DATA/users/$username/document.db` for the database when trying to login/sync, and if the file/folder doesn't exist, fall back on the default `$TRILLIUM_DATA/document.db`, this would allow us to have a single server to sync the db of everyone's local trillium db. As it stands, it's not practical for me to start a new server, and define a new subdomain/etc.. for each person in the team to sync their local trillium to the server, for example. But having a single server able to sync multiple trilliums/users and store each in a separate subfolder which be a huge help.

A third step (and one that we might not get to, but which I think is the pandora's box you are worried about) would be to have some way of sharing notes between accounts. At this time, I think this isn't necessary for my purposes, but eventually, it would be nice to be able to select a note from one's account and share it with other accounts. I have some ideas on how to achieve this in a non-disruptive manner, but it's not worth talking about at this time.

Note: this doesn't tackle any possible use case of "multiple users editing the same note at the same time" because that's not really the point for me, and it would add a lot to the complexity.

Thank you for reading and if you have any suggestions, thoughts or other ideas you'd like to share, we'd appreciate it!

**Comment by zadam** [comment]
> The very first use case to implement would be to still have one account with be ability to lock notes/subtree with specific passwords. 

That sounds relatively easy to do. But to be secure, you need to disable all custom scripting (e.g. render notes, scripts running after loading of the frontend) since such script would wait until you unlock the notes, read them and perhaps send them somewhere.

> I think the next step might be to have Trillium automatically try to open $TRILLUM_DATA/users/$username/document.db for the database when trying to login/sync

That does not sound good. Trilium opens the DB file at the startup and actually makes quite a lot of things with it - it e.g. reads all metadata (e.g. tree structure) in memory to speed up things, it will check for DB consistency, fill missing records etc. A single Trilium process can operate only on a single document.db, and you would have to rewrite a big part of the backend application to change that.

**Comment by kakaroto** [comment]
Thanks for that feedback. I did see the support for scripting as one issue with multi user environments, and I think having those features disabled automatically when multi-user support is enabled would be necessary, (and an easy way to avoid trouble, rather than trying to make it secure with those features).

I know Trilium works on a single document.db, but having it support more than one *should* be possible.. I don't know how it does things and the complexity yet of the backend, so I'm just looking at it with an external/abstract view right now. Ideally, it would be able to close/unload the current document.db and switch to another one. I think the biggest issue and the one you're talking about is that it wouldn't allow multiple users to concurrently log in, because only one db can be opened at a time, which isn't something I thought of. 
I'm currently only thinking about the server being used for syncing of the data from multiple different users... I haven't thought of users logging in directly into the server itself. I'll have to think about whether or not it's something I'd be willing to sacrifice.

Thanks again for the feedback! I might not have time to look into the Trilium work for the next couple of months, but once I have a clearer idea of how to move forward, I'll let you know (if you're interested in the design decision/progress).

**Comment by meichthys** [comment]
@kakaroto & @Varstahl 
If/When you start your changes, could you report back here with a link to your repository?

**Comment by Varstahl** [comment]
Yes, absolutely. We had other urgent priorities that caught us off guard, but when we start working on it it will be forked here on GitHub, no problems.

**Comment by BrettBass** [comment]
Any update @Varstahl

**Comment by kakaroto** [comment]
> Any update @Varstahl

Hey, sorry. We've had other more urgent work recently and our priorities shifted, we haven't started working on this quite yet.

**Comment by serrapa** [comment]
Any update  @Varstahl . Sorry If i ask you about updates, but sounds a great project what you all want to do!

**Comment by kakaroto** [comment]
> Any update @Varstahl . Sorry If i ask you about updates, but sounds a great project what you all want to do!

Sorry, unfortunately, the project had to be dropped as it was too complicated to try and modify TN to do what we needed it to. We've seen switch to a different software stack for our notes/knowledge base

**Comment by FedorAdo** [comment]
> Sorry, unfortunately, the project had to be dropped as it was too complicated to try and modify TN to do what we needed it to. We've seen switch to a different software stack for our notes/knowledge base

Can I find out what you switched to?

**Comment by kakaroto** [comment]
> Can I find out what you switched to?

We've switched to [Outline](https://github.com/outline/outline). 
I wrote a script to import the exported data from TN into Outline, preserving the structure and attachments.
Note that the script was made for our single use, I used it to import our content once and never re-used it, so it may be buggy, incomplete or outdated, etc.. as it was never meant for public consumption

I'm hoping that this isn't breaking any kind of netiquette (I have no vested interest in Outline, so this isn't spamming, but it's still referencing a different project, and some might take offense with that, even if it's open source).
If it's an issue for @zadam, then I apologize in advance, and feel free to delete my comment, I would fully understand.

Here's the script (you'd need to `npm install argparse fs-extra form-data node-fetch` if I remember correctly)
```js
const { ArgumentParser } = require('argparse')
const fs = require("fs-extra");
const path = require("path");
const FormData = require('form-data');
const fetch = require('node-fetch');

async function post(endpoint, data, {outlineUrl, token, method="POST"}={}) {
    const response = await fetch(`${outlineUrl}/api/${endpoint}`, {
        method: "POST",
        headers: {
            Accept: "application/json",
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`
        },
        body: JSON.stringify(data)
    })

    const body = await response.json();
    if (!body.ok) {
        console.error(`Error in ${endpoint} API : ${body.error}`);
        return null;
    }
    return body.data;
}

async function upload(url, form, file) {
    const formData = new FormData();
    for (const key in form) {
        formData.append(key, form[key]);
    }
    formData.append("file", file);
    const response = await fetch(url, {
        method: "POST",
        body: formData
    })

    if (!response.ok) {
        console.error(`Error in upload : ${response.status}`);
        return false;
    }
    return true;
}

async function importNote(collection, parent, meta, dataPath, outlineUrl, token) {
    let text = undefined;
    if (meta.dataFileName) {
        if (meta.format === "markdown") {
            text = await fs.readFile(path.resolve(dataPath, meta.dataFileName), {encoding: 'utf-8'});
        } else {
            const attachmentBlob = await fs.readFile(path.resolve(dataPath, meta.dataFileName));
            const originalFileName = meta.attributes.find(a => a.name === "originalFileName");
            const filename = originalFileName ? originalFileName.value : meta.dataFileName;
            const attachment = await post("attachments.create", {
                name: filename,
                contentType: meta.mime,
                size: attachmentBlob.length
            }, {outlineUrl, token});
            await upload(attachment.uploadUrl, attachment.form, attachmentBlob);
            const isImage = meta.mime.startsWith("image");
            text = `${isImage ? '!' : ''}[${isImage ? "": `${filename} ${attachmentBlob.length}`}}](/api/attachments.redirect?id=${attachment.attachment.id})`;            
        }
    }
    const note = await post("documents.create", {
        title: meta.title,
        collectionId: collection.id,
        text: text,
        parentDocumentId: parent ? parent.id : undefined,
        publish: true
    }, {outlineUrl, token});

    if (meta.children && meta.children.length > 0) {
        // Add them in reverse order since outline orders new documents to the top of the tree
        for (let i = meta.children.length - 1; i >= 0; i--) {
            const child = meta.children[i];
            const childPath = path.resolve(dataPath, meta.dirFileName);
            await importNote(collection, note, child, childPath, outlineUrl, token);
        }
    }
    return note;
}

async function importCollection(meta, dataPath, outlineUrl, token) {
    const collection = await post("collections.create", {
        name: meta.title,
        private: true,
        sharing: false
    }, {outlineUrl, token});
    // If collection has data, import it as a separate note but with no children
    if (meta.dataFileName) {
        await importNote(collection, null, {...meta, children: []}, dataPath, outlineUrl, token);
    }
    if (meta.children && meta.children.length > 0) {
        // Add them in reverse order since outline orders new documents to the top of the tree
        for (let i = meta.children.length - 1; i >= 0; i--) {
            const child = meta.children[i];
            const childPath = path.resolve(dataPath, meta.dirFileName);
            await importNote(collection, null, child, childPath, outlineUrl, token);
        }
    }
}

async function importTN(root, outlineUrl, token) {
    const meta = await fs.readJSON(path.resolve(root, "!!!meta.json"));
    for (const file of meta.files) {
        if (file.noteId === "root") {
            // Add them in reverse order since outline orders new documents to the top of the tree
            for (let i = file.children.length - 1; i >= 0; i--) {
                const child = file.children[i];
                const childPath = path.resolve(root, file.dirFileName);
                await importCollection(child, childPath, outlineUrl, token);
            }
        } else {
            console.error("Non root import!");
            break;
        }
    }
}

async function init() {
    const parser = new ArgumentParser({
        add_help: true,
        description: 'Import Trilium Notes export to Outline'
    });
    parser.add_argument("root", { help: 'Root folder to import' });
    parser.add_argument("url", { help: 'The URL of the outline installation', default: "https://app.getoutline.com" });
    parser.add_argument("token", { help: 'API token' });
    const args = parser.parse_args();

    await importTN(args.root, args.url, args.token);
    process.exit(0);
}

init();

```

**Comment by FedorAdo** [comment]
> We've switched to [Outline](https://github.com/outline/outline).

Thank you.
